provisioning/hare_setup.py

```diff
@@ -0,0 +1,301 @@
+            exit(-1)
+
+class SupportBundle(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            # Default target directory is /tmp/hare
+            if os.system('hctl reportbug') != 0:
+                logging.error('Failed to generate support bundle')
+                exit(-1)
+        except Exception as error:
+            logging.error('Error while generating support bundle (%s)',
+                          error)
+            exit(-1)
+
+class Cleanup(argparse.Action):
+    def __call__(self, parser, namespace, values, option_string=None):
+        try:
+            exit(0)
+        except Exception as error:
+            logging.error('Failed to perform cleanup (%s)',
+                          error)
+            exit(-1)
+
+def checkRpm(rpm_name):
+    rpm_list = subprocess.Popen(["rpm", "-qa"],
+                                stdin=subprocess.PIPE,
+                                stdout=subprocess.PIPE,
+                                stderr=subprocess.PIPE,
+                                env={},
+                                encoding='utf8')
+    rpm_search = subprocess.Popen(["grep", "-q", rpm_name],
+                                  stdin=rpm_list.stdout,
+                                  stdout=subprocess.PIPE,
+                                  stderr=subprocess.PIPE,
+                                  env={},
+                                  encoding='utf8')
+    out, err = rpm_search.communicate()
+    logging.debug("Output: {}".format(out))
+    logging.debug("stderr: {}".format(err))
+
+    return rpm_search.returncode
+
+
+def is_cluster_running() -> bool:
+    return os.system('hctl status >/dev/null') == 0
+
+
+def bootstrap_cluster():
+    return os.system('hctl bootstrap --mkfs /var/lib/hare/cluster.yaml')
+
+
+def shutdown_cluster():
+    return os.system('hctl shutdown')
+
+
+def list2dict(nodes_data_hctl: list) -> dict:
+    node_info_dict = {}
+    for node in nodes_data_hctl:
+        node_svc_info = {}
+        for service in node['svcs']:
+            node_svc_info[service['name']] = []
+            if (service['status'] == 'started'):
+                node_svc_info[service['name']].append(service['status'])
+        node_info_dict[node['name']] = node_svc_info
+
+    return node_info_dict
+
+
+def check_cluster_status():
+    cluster_desc = None
+    with open("/var/lib/hare/cluster.yaml", 'r') as stream:
+        cluster_desc = yaml.safe_load(stream)
+    cmd = ['hctl', 'status', '--json']
+    cluster_info = json.loads(execute(cmd))
+    nodes_data_hctl = cluster_info['nodes']
+
+    node_info_dict = list2dict(nodes_data_hctl)
+    for node in cluster_desc['nodes']:
+        m0client_s3_cnt = node['m0_clients']['s3']
+        m0ds = node.get('m0_servers', [])
+        for m0d in m0ds:
+            ios_cnt = 0
+            if ('runs_confd' in m0d.keys() and
+                (node_info_dict[node['hostname']]['confd'][0] != 'started')):
```

Comment:
Does this code work with the JSON structure like this?

```json
 "nodes": [
    {
      "name": "localhost",
      "svcs": [
        {
          "name": "hax",
          "fid": "0x7200000000000001:0x6",
          "ep": "192.168.6.214@tcp:12345:1:1",
          "status": "started"
        },
        {
          "name": "confd",
          "fid": "0x7200000000000001:0x9",
          "ep": "192.168.6.214@tcp:12345:2:1",
          "status": "started"
        },
        {
          "name": "ioservice",
          "fid": "0x7200000000000001:0xc",
          "ep": "192.168.6.214@tcp:12345:2:2",
          "status": "started"
        },
        {
          "name": "m0_client",
          "fid": "0x7200000000000001:0x28",
          "ep": "192.168.6.214@tcp:12345:4:1",
          "status": "unknown"
        },
        {
          "name": "m0_client",
          "fid": "0x7200000000000001:0x2b",
          "ep": "192.168.6.214@tcp:12345:4:2",
          "status": "unknown"
        }
      ]
    }
  ]
```
If yes, I don't quite understand why there is 0 index in your expression `node_info_dict[node['hostname']]['confd'][0] != 'started'`

Reply:
node_info_dict[node['hostname']]['confd'] is a list. Actually node_info_dict[node['hostname']][<service name>] is list and in case of 'ioservices' it will contain multiple entries of status(started).
In case of confd only 1 entry will be there thats why node_info_dict[node['hostname']]['confd'][0] != 'started'